frame/im-online/src/lib.rs

```diff
@@ -315,7 +315,22 @@ decl_module! {
 
 		fn deposit_event() = default;
 
-		#[weight = MINIMUM_WEIGHT]
+		// NOTE: the weight include cost of validate_unsigned as it is part of the cost to import
+		// block with such an extrinsic.
+		/// # <weight>
+		/// - Complexity: `O(K + E)` where K is length of `Keys` and E is length of
+		///   `Heartbeat.network_state.external_address`
+		///
+		///   - `O(K)`: decoding of length `K`
+		///   - `O(E)`: decoding/encoding of length `E`
+		/// - DbReads: pallet_session `Validators`, pallet_session `CurrentIndex`, `Keys`,
+		///   `ReceivedHeartbeats`
+		/// - DbWrites: `ReceivedHeartbeats`
+		/// # </weight>
+		#[weight = 310_000_000
+			+ 750_000 * Keys::<T>::decode_len().unwrap_or(0) as Weight
+			+ 1_200_000 * heartbeat.network_state.external_addresses.len() as Weight
+			+ T::DbWeight::get().reads_writes(4, 1)]
 		fn heartbeat(
```

Comment:
calling `Keys::decode_len` is fine to me but should be double-checked.

AFAIU calling a storage in the computation of the weight here is OK because validate_unsigned has already put `Keys` into the cache already.

If this is an issue we can put this into an argument and check that the argument is correct.